### ISSUE TITLE
HttpClient/Handler: Add support for client-side cache-control

### ISSUE THREAD

**Issue by dotMorten** [dotnet/runtime#1384]
In .NET Framework it was possible to specify that web requests should use the shared browser cache by setting:
`myWebRequestHandler.CachePolicy = new System.Net.Cache.RequestCachePolicy(System.Net.Cache.RequestCacheLevel.Default);`

This allows repeated web requests to either be pulled from the local browser cache (if not expired) or do much faster ETag 304 requests/response. This is really really useful for client apps that frequently check for status updates in a resource, image tile requests etc, and can reduce the network overhead tremendously.

For example I made a (very) quick and dirty cache implementation showing a 33x perf increase on repeated requests when implementing cache-control max-age and/or etag:

|                Method |      Mean |     Error |    StdDev |    Median | Ratio | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------------- |----------:|----------:|----------:|----------:|------:|------:|------:|------:|----------:|
| UseSocketsHttpHandler | 48.710 ms | 0.4979 ms | 0.4414 ms | 48.694 ms |  1.00 |     - |     - |     - | 243.08 KB |
|  UseHttpClientHandler | 47.781 ms | 0.5331 ms | 0.4451 ms | 47.884 ms |  0.98 |     - |     - |     - | 242.25 KB |
|   UseCacheHttpHandler |  1.252 ms | 0.0665 ms | 0.1907 ms |  1.199 ms |  0.03 |     - |     - |     - | 247.48 KB |

Test was done on an above-average high-speed internet connection. Effect would of course be A LOT bigger on slower networks.

However building a proper caching mechanism that deals with shared vs private cache, cache expiration, pruning (when it grows too large), lock contention for parallel requests, re-using the shared OS browser cache (like what wininet offers), etc is no trivial task. And even so, not having this at a platform level means there's no way to make caching possible with 3rd party libraries, because it currently has to be done manually at a very low level of the request.

/CC @karelz

**Comment by karelz** [comment]
@dotMorten given that we do not have any tight integration with wininet in SocketsHttpHandler, there is no benefit to have it in the platform. 3rd party library will have same chance to implement it as us.
Given that there is ton of policy around caching (e.g. store it on file system, or in memory, how much?), we do not think it should be part of SocketsHttpHandler inbox.

As such, I recommend to close the issue as Won't Fix. What do you think?

**Comment by dotMorten** [comment]
You know what I would think :-) Caching is an integral part of client apps - it needs an OOB option. We have it in .NET Framework, and we have it in UWP for good reasons. We don't however have it anywhere in .NET Core.

First of all, I don't think it's fair putting this on anyone to build themselves. I've only just completed building something that is only just good enough for my specific use-case, and it was a significant amount of work.

>  3rd party library will have same chance to implement it as us.

I don't see how? There are no common extensibility points that a 3rd party library can use that would turn it on for all dependencies. You'd have to build your own handler (or delegate handler around the sockets handler).

> Given that there is ton of policy around caching

I've been through the RFCs around this a lot lately, and there's not a ton. It's actually a quite short list. See https://tools.ietf.org/html/rfc7234#section-3
```
 A cache MUST NOT store a response to any request, unless:

   o  The request method is understood by the cache and defined as being
      cacheable, and

   o  the response status code is understood by the cache, and

   o  the "no-store" cache directive (see Section 5.2) does not appear
      in request or response header fields, and

   o  the "private" response directive (see Section 5.2.2.6) does not
      appear in the response, if the cache is shared, and

   o  the Authorization header field (see Section 4.2 of [RFC7235]) does
      not appear in the request, if the cache is shared, unless the
      response explicitly allows it (see Section 3.2), and

   o  the response either:

      *  contains an Expires header field (see Section 5.3), or

      *  contains a max-age response directive (see Section 5.2.2.8), or

      *  contains a s-maxage response directive (see Section 5.2.2.9)
         and the cache is shared, or

      *  contains a Cache Control Extension (see Section 5.2.3) that
         allows it to be cached, or

      *  has a status code that is defined as cacheable by default (see
         Section 4.2.2), or
```
Or to put it more simple: It basically says to only cache `GET`/`HEAD` and only if there's no `Authorization`* header, it doesn't specify `NoStore` or `Private`*, there's an etag or expiration and you get a `200 OK`.

Second, it should of course be opt-in, and not on by default (just like it is on netfx).

* You could store Auth and Private, but the simplest / safest solution is just to no do that, which is also how NETFX behaves.

**Comment by thargol1** [comment]
Perhaps caching can be implemented as on opt in on `HttpClientFactory`?

Some like this in startup.cs
```csharp
services.AddHttpClient().AddCacheHandler(options =>
{
     // what ever you guys come up with, but maybe someway to inject an IMemoryCache here
});
```
You can than use the cache handler when it's allowed to do, as specified by the cache-headers of the returning response, or leave it up to the programmer.

**Comment by karelz** [comment]
Triage: We believe this kind of opinionated policy (esp. potential wininet integration) does not belong into vanila `HttpClient`, but it makes sense to keep it in `HttpClientFactory`.

**Comment by johnkors** [comment]
>It basically says to only cache GET/HEAD and only if there's no Authorization* header, it doesn't specify NoStore or Private*, there's an etag or expiration and you get a 200 OK.

I have an odd scenario where I'd like caching even though a _Authorization_ header is present, so please make it configurable if you're considering implementing the rules mentioned by @dotMorten . I don't mind sensible defaults, as long as it's allowing for configuration.

**Comment by StevenLiekens** [comment]
I believe this should be in SocketsHttpHandler because it can mutate request and response headers that participate in cache keys.

E.g. AutomaticDecompression

``` csharp
services.AddHttpClient("demo")
    .ConfigurePrimaryHttpMessageHandler(() => new SocketsHttpHandler
    {
        AutomaticDecompression = DecompressionMethods.GZip
    });
```

A side-effect of this configuration is that all requests are sent with `Accept-Encoding: gzip`, but that is not observable in a `DelegatingHandler` _before_ sending the request. A well-behaved server will respond with `Vary: Accept-Encoding` but again there is no way to check it when deciding whether to use a cached response, without sending the request.

You could argue that you can just ignore the `Vary: Accept-Encoding` since the automatic decompression takes care of decoding the stream... but that's probably a sin against rfc7234 and I'd prefer caching the gzipped content anyway.

This is just one example, I think `PreAuthenticate` has similar problems.

**Comment by dotMorten** [comment]
@StevenLiekens I implemented my own caching at this level as well, and didn't worry too much about the gzip - I don't see how that really changes the cache expiration, etags etc. You'd get the same thing in the end anyway

**Comment by StevenLiekens** [comment]
I would just not recommend caching a projection of what is actually returned by the server, especially when the request is mutated between a cache miss and sending it to the server. I think my case is clearer for the PreAuthenticate feature.

**Comment by StevenLiekens** [comment]
In other words

![image](https://user-images.githubusercontent.com/1583241/124660594-b35bb000-dea6-11eb-96b6-546fa4b19f5d.png)

**Comment by SteveDunn** [comment]
Has there been any traction on this?  Is CacheCow still the way to go currently?

**Comment by StevenLiekens** [comment]
There is still a gap that needs to be filled. CacheCow isn't for everyone, it's most useful when you control both the client and the server.  I did diagonal scan of the code and it doesn't seem to implement all the nitty gritty details of RFC 7234.

I wish MS would come forward with a WebCacheHandler  with pluggable storage (DistributedCache?) that implements the full RFC. They already implemented RFC 7230-7233 so why stop at 7234?

**Comment by SteveDunn** [comment]
@StevenLiekens - that's a shame.  This stuff is certainly a lot more dificcult since .NET Core / 5.

**Comment by StevenLiekens** [comment]
FWIW I'm now doing the same thing as Morten, writing my own cache handler. It's feasible for me because I'm doing this outside of normal work hours for a pet projects... but you can't honestly expect someone to implement the entire web caching draft as part of their job building ecommerce apps. https://github.com/StevenLiekens/gw2sdk/blob/ad310e5b98d4db12229f831144617edeb6c4aeab/GW2SDK/Http/Caching/CachingHttpHandler.cs

I somewhat agree with Karel that storage policies shouldn't become a concern for the SocketsHttpHandler, but then neither is automatic decompression (content negotiation), automatic preauthentication and automatic following redirects. I suspect these things carried over from old HttpClientHandler code, but now they are just in my way. My handler only works reliably when you disable all of these features.

Long story short, I think caching behavior should be a core feature, whether or not it is integrated in SocketsHttpHandler or implemented as a stand-alone handler. It doesn't have to be opinionated code. Storage APIs can be injected through DI with HttpClientFactory. Storage policies can be implemented in user code if the caching handler raises an event whenever it makes a decision about a request (`RequestForwarded` / `ResponseReused` / `ResponseValidated`).

**Comment by geoffkizer** [comment]
> I somewhat agree with Karel that storage policies shouldn't become a concern for the SocketsHttpHandler, but then neither is automatic decompression (content negotiation), automatic preauthentication and automatic following redirects. I suspect these things carried over from old HttpClientHandler code, but now they are just in my way. My handler only works reliably when you disable all of these features.

This is an interesting point, and worth considering on its own, regardless of whether we end up building caching into SocketsHttpHandler or not.

Compression, authentication, redirect handling, and caching are all capabilities that logically are layered on top of the core protocol handling. And there are other capabilities too that you could imagine being layered in similar ways, such as retry policy, routing, load balancing, etc.

But to work properly, these capabilities typically need to be ordered properly. For example, just focusing on the capabilities built in to SocketsHttpHandler: Redirect handling is logically layered "above" authentication, because a redirect can change the target server and thus change how authentication should be performed.

We have currently built a set of capabilities into SocketsHttpHandler (decompression, authentication, redirects, cookie handling) and we execute them in the appropriate order for this set of capabilities. And in fact, most of these are implemented similar to a delegating handler internally, though they are not exposed publicly as such.

What we don't allow you to do is insert your own delegating handler *between* the built-in layered capabilities like decompression, redirect handling, etc. The result is that if you want to implement other layered capabilities, like caching, you may have to entirely forgo the built-in capabilities in order to get an ordering that isn't broken. And if you want to replace one of the built-in capabilities with your own custom logic -- e.g. control redirect handling or provide your own authentication -- you are similarly impacted.

Long term we should consider breaking these layered capabilities out into separate delegating handlers so that they can be reused in situations like this.

**Comment by jacobmohl** [comment]
Just to chime in on this aswell. It should be in the ASP.Net or .Net package in some way, Either a middleware in ASP.Net or a configurable part of HttpClient(Factory).

Its a standard with (RFC) so there should not be that much to discus beside storage mechanism.

**Comment by dotMorten** [comment]
It definitely don’t think it should be in ASP.NET. The biggest use for this would be client applications

**Comment by brantburnett** [comment]
I recognize that caching is complicated, and implementing it is a big ask. That said, I think one of the key takeaways from this thread is that it currently isn't possible to effectively implement caching in `HttpClient`, at least not without rewriting a whole mass of unrelated code or accepting some odd limitations. This means that even if I wanted to create a NuGet package that extends `HttpClient` with caching, it would be, at best, a mess.

Could we not consider a more manageable middle road? Instead of implementing a complete caching option in .NET, provide an extension point where a cache may be plugged into SocketsHttpHandler at the correct point in the stack (I'd assume after all other "middleware" like decompression, redirect, etc). This way 3rd-party authors could implement caching (hopefully as an open-source NuGet for everyone to use). And if the .NET team ever decides to in-box a caching solution it would just be an in-box implementation to plugin to the extension point.

Possible APIs could look something like this:

```c#
namespace System.Net.Http;

// The next parameter is the method to forward the request to on cache miss or refresh, potentially modified with `If-Modified-Since`, etc.
public delegate ValueTask<HttpResponseMessage> HttpCacheHandler(
    HttpRequestMessage request, 
    Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> next, 
    CancellationToken cancellationToken);

public partial class SocketsHttpHandler
{
    public HttpCacheHandler? Cache { get; set; }
}
```

Or:

```c#
namespace System.Net.Http;

public interface IHttpClientCache
{
    // The next parameter is the method to forward the request to on cache miss or refresh, potentially modified with `If-Modified-Since`, etc.
    ValueTask<HttpResponseMessage> GetOrSendAsync(
        HttpRequestMessage request, 
        Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> next, 
        CancellationToken cancellationToken);

    // Other members here, perhaps Clear or a simple TryGet that doesn't forward
}

public partial class SocketsHttpHandler
{
    public IHttpClientCache? Cache { get; set; }
}
```

**Comment by dotMorten** [comment]
@brantburnett I'm not sure you say it can't be implemented today. We did it. You just need to overwrite sendasync. 
However the main problem with both your suggestion and our solution is it only works for the one 3rd party library that chose to do this. There's no way to have an entire app and all it's libraries to all get caching enabled together.

**Comment by brantburnett** [comment]
> I'm not sure why you say it can't be implemented today. We did it. You just need to overwrite sendasync.

I would disagree with this statement (in part). I agree it is possible to implement a caching layer. However, I don't believe it is possible to *generically* implement an RFC-compliant caching layer that correctly handles `Accept-Encoding` and caches redirects, etc,  without making some significant compromises. 

A non-compliant caching layer is primarily useful for cases where you control the server and are therefore certain it will only implement certain patterns. To create a reusable NuGet package (which I'd like to do) that is RFC-compliant and broadly compatible I would probably do something like this today:

1. Implement as a `DelegatingHandler`
2. Not cache redirects
3. Instruct the consumer to never enable `AutomaticDecompression` on `SocketsHttpHandler` or `HttpClientHandler`
4. Either not support automatic decompression at all (bad for perf), or reimplement automatic decompression in the caching handler (bloat and lots of room for bugs and performance missteps)

For a generic NuGet package, number 3 is especially problematic since it represents a "pit of failure" for the consumer. I'd much prefer something that offers a "pit of success".

I believe the only way to implement a consistent, easy-to-consume, generic, and RFC-compliant caching layer is to insert it here between the `DiagnosticsHandler` and the `RedirectHandler` as a `HttpMessageHandlerStage`, which is currently internal: https://github.com/dotnet/runtime/blob/2c471afb09122433dc8137a9827e47ff8cedd6ff/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L538-L560

My proposals above are simply ways to plug into the stack at that point without exposing too many internals.

**Comment by dotMorten** [comment]
We support all those things. Granted it took many releases to get completely right and handle all edge cases and ensure signed in users won't share cache with other users etc. Not something the average developer has the time and resources to do and why I logged this issue in the first place (in addition to getting a way that will affect all handler instances). Wish we could open-source the bits but it was a significant investment for us and a compete feature so doubt we'll be able to. I don't get your step 3 and 4. The cache doesn't need to be compressed. In fact that just means you'll be wasting time decompressing each time. The disk space savings just isn't worth it.

**Comment by brantburnett** [comment]
> I don't get your step 3 and 4. The cache doesn't need to be compressed. In fact that just means you'll be wasting time decompressing each time. The disk space savings just isn't worth it.

See the explanation above for varying caching by `Accept-Encoding`: https://github.com/dotnet/runtime/issues/1384#issuecomment-873611501

Not varying based on the encoding requested could cause misbehavior per the RFC. In theory, a well-behaved server would always serve the same response with different compression, but that may not be the case. So a properly implemented cache should cache the compressed variant and follow the returned `Vary` header to vary the cache based on request headers such as `Accept-Encoding`. A generic, reusable implementation should avoid such pitfalls.

Additionally, you're assuming that the cache is on disk. Implementations may prefer to cache in memory and the reduced memory utilization may be worthwhile. Or, the consumer may prefer a distributed cache such as Redis or Couchbase, where network latency is a factor and smaller compressed packets are advantageous. Even on disk, depending on disk I/O speed, available CPU cycles, and compression ratio, decompressing each time may provide more throughput in some cases.

**Comment by dotMorten** [comment]
You can still handle the vary-by headers even if you uncompress it. If I were to query the same endpoint with two different encodings, and the server says the response varies by encoding, then they would be two separate cache entries.

**Comment by sliekens** [comment]
I abandoned my caching side-quest because I feel uncomfortable caching a projection of the response, rather than the actual response. It doesn't sit right with me that when the requester does `Accept-Encoding: gzip`, the cache might reply with an already-decoded response. This might not even work between different instances of `HttpClient` that share a cache, but have different settings for `AutomaticDecompression`. I suspect the same problems exist with `AllowAutoRedirect` and `PreAuthenticate`.

**Comment by sliekens** [comment]
Code to demonstrate the above: https://dotnetfiddle.net/w3XmcW

```csharp
var http = new HttpClient(new SocketsHttpHandler { AutomaticDecompression = System.Net.DecompressionMethods.GZip });

Console.WriteLine("Request before sending (this is what you use to search your cache):\n");
var request = new HttpRequestMessage(HttpMethod.Get, "https://github.com/dotnet/runtime/issues/1384");
LogRequestMessage(request);

Console.WriteLine("Response from server:\n");
using var response = http.Send(request);
LogResponseMessage(response);

Console.WriteLine("Request after sending (this is what you use to build the cache key):\n");
LogRequestMessage(request);
```

Output:

```text
Request before sending (this is what you use to search your cache):

    GET https://github.com/dotnet/runtime/issues/1384 HTTP/1.1

Response from server:

    HTTP/1.1 200 OK
    Server: GitHub.com
    Date: Tue, 29 Oct 2024 23:50:41 GMT
    Cache-Control: must-revalidate, max-age=0, private
    Content-Security-Policy: default-src 'none'; base-uri 'self'; child-src github.com/assets-cdn/worker/ github.com/webpack/ github.com/assets/ gist.github.com/assets-cdn/worker/; connect-src 'self' uploads.github.com www.githubstatus.com collector.github.com raw.githubusercontent.com api.github.com github-cloud.s3.amazonaws.com github-production-repository-file-5c1aeb.s3.amazonaws.com github-production-upload-manifest-file-7fdce7.s3.amazonaws.com github-production-user-asset-6210df.s3.amazonaws.com *.rel.tunnels.api.visualstudio.com wss://*.rel.tunnels.api.visualstudio.com objects-origin.githubusercontent.com copilot-proxy.githubusercontent.com proxy.individual.githubcopilot.com proxy.business.githubcopilot.com proxy.enterprise.githubcopilot.com *.actions.githubusercontent.com wss://*.actions.githubusercontent.com productionresultssa0.blob.core.windows.net/ productionresultssa1.blob.core.windows.net/ productionresultssa2.blob.core.windows.net/ productionresultssa3.blob.core.windows.net/ productionresultssa4.blob.core.windows.net/ productionresultssa5.blob.core.windows.net/ productionresultssa6.blob.core.windows.net/ productionresultssa7.blob.core.windows.net/ productionresultssa8.blob.core.windows.net/ productionresultssa9.blob.core.windows.net/ productionresultssa10.blob.core.windows.net/ productionresultssa11.blob.core.windows.net/ productionresultssa12.blob.core.windows.net/ productionresultssa13.blob.core.windows.net/ productionresultssa14.blob.core.windows.net/ productionresultssa15.blob.core.windows.net/ productionresultssa16.blob.core.windows.net/ productionresultssa17.blob.core.windows.net/ productionresultssa18.blob.core.windows.net/ productionresultssa19.blob.core.windows.net/ github-production-repository-image-32fea6.s3.amazonaws.com github-production-release-asset-2e65be.s3.amazonaws.com insights.github.com wss://alive.github.com api.githubcopilot.com api.individual.githubcopilot.com api.business.githubcopilot.com api.enterprise.githubcopilot.com; font-src github.githubassets.com; form-action 'self' github.com gist.github.com copilot-workspace.githubnext.com objects-origin.githubusercontent.com; frame-ancestors 'none'; frame-src viewscreen.githubusercontent.com notebooks.githubusercontent.com; img-src 'self' data: blob: github.githubassets.com media.githubusercontent.com camo.githubusercontent.com identicons.github.com avatars.githubusercontent.com private-avatars.githubusercontent.com github-cloud.s3.amazonaws.com objects.githubusercontent.com secured-user-images.githubusercontent.com/ user-images.githubusercontent.com/ private-user-images.githubusercontent.com opengraph.githubassets.com github-production-user-asset-6210df.s3.amazonaws.com customer-stories-feed.github.com spotlights-feed.github.com objects-origin.githubusercontent.com *.githubusercontent.com; manifest-src 'self'; media-src github.com user-images.githubusercontent.com/ secured-user-images.githubusercontent.com/ private-user-images.githubusercontent.com github-production-user-asset-6210df.s3.amazonaws.com gist.github.com; script-src github.githubassets.com; style-src 'unsafe-inline' github.githubassets.com; upgrade-insecure-requests; worker-src github.com/assets-cdn/worker/ github.com/webpack/ github.com/assets/ gist.github.com/assets-cdn/worker/
    Link: </_graphql?body=%7B%22query%22%3A%22bc81d213bef33096858522d757775720%22%2C%22variables%22%3A%7B%22markAsRead%22%3Atrue%2C%22number%22%3A1384%2C%22owner%22%3A%22dotnet%22%2C%22repo%22%3A%22runtime%22%2C%22useNewTimeline%22%3Afalse%7D%7D>; rel=preload; as=fetch; crossorigin=use-credentials
    Referrer-Policy: no-referrer-when-downgrade
    Server-Timing: issue_layout-fragment;desc="issue_layout fragment";dur=179.754858,issue_conversation_content-fragment;desc="issue_conversation_content fragment";dur=476.064631,issue_conversation_sidebar-fragment;desc="issue_conversation_sidebar fragment";dur=58.432276,nginx;desc="NGINX";dur=1.092435,glb;desc="GLB";dur=34.215366
    Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
    Vary: X-PJAX, X-PJAX-Container, Turbo-Visit, Turbo-Frame, Accept, Accept-Encoding, Accept, X-Requested-With
    X-Content-Type-Options: nosniff
    X-Frame-Options: deny
    x-voltron-version: 69a2227
    X-XSS-Protection: 0
    Set-Cookie: _gh_sess=rEEKr1C0kuHmy6BSABQf6MlvmlW8V5xMn5JHjfh%2F7FKeuQWpYPOdk23CoS5WO4c7SpBfH4TjOYWCjpp0XVVV45NHeyN47Dxvrc%2FC3WproHUaRIR8wFhAKYcW6LNCpiHDq1ATS3LxkAL7JL7DJ0fjjcK2jsyek408IF1UGKzAo5IXu3Sk3ut1pSI5%2BvhjooIjIoN%2B6vrhzDJKkW1dHOph6zG3tweDmrbVp5jvLkVvv316BtwM%2FDKozub7MqdK6q0gtkIYQJsZsdX2wMz0To1ayw%3D%3D--gkRLZk%2BoYyPu2S4p--5eSeiNSmKRW9czHawkgO2Q%3D%3D; Path=/; HttpOnly; Secure; SameSite=Lax, _octo=GH1.1.1850364850.1730245855; Path=/; Domain=github.com; Expires=Wed, 29 Oct 2025 23:50:55 GMT; Secure; SameSite=Lax, logged_in=no; Path=/; Domain=github.com; Expires=Wed, 29 Oct 2025 23:50:55 GMT; HttpOnly; Secure; SameSite=Lax
    Accept-Ranges: bytes
    Transfer-Encoding: chunked
    X-GitHub-Request-Id: 8E00:3D311:193DF0B0:19B13B44:672174DF
    Content-Type: text/html; charset=utf-8

Request after sending (this is what you use to build the cache key):

    GET https://github.com/dotnet/runtime/issues/1384 HTTP/1.1
    Accept-Encoding: gzip
```
